python/tskit/tables.py

```diff
@@ -3020,6 +3026,99 @@ def union(
                 record=json.dumps(provenance.get_provenance_dict(parameters))
             )
 
+    def time_join(self, recent, time, continue_nodes=None):
+        ancient_ts = self.tree_sequence()
+        recent_ts = recent.tree_sequence()
+        if continue_nodes is None:
+            continue_nodes = ancient_ts.samples()
+
+        new_tables = recent_ts.tables
+        new_nodes = np.empty(0, dtype=np.uint32)
+        for t in recent_ts.trees():
+            for root_id in t.roots:
+                root = recent_ts.node(root_id)
+                if root.time == time:
+                    new_nodes = np.append(new_nodes, root_id)
+                elif root.time < time:
+                    n = new_tables.nodes.add_row(time=time)
+                    new_tables.edges.add_row(
+                        left=t.interval.left,
+                        right=t.interval.right,
+                        parent=n,
+                        child=root_id,
+                    )
+                    new_nodes = np.append(new_nodes, n)
+                else:
+                    raise RuntimeError(
+                        f"Root found older than join time (age = {root.time})"
+                    )
+        recent_ts = new_tables.tree_sequence()
+
+        if len(new_nodes) > len(continue_nodes):
+            raise RuntimeError(
+                f"""More lineages present at the longest-ago end of `recent_ts` \
+                than there are nodes in `continue_nodes`. \
+                Setting sample_size={len(continue_nodes)} or leaving it to default \
+                will ensure compliance."""
+            )
+
+        node_map = np.repeat(tskit.NULL, recent_ts.num_nodes)
+
+        node_map[new_nodes] = np.random.choice(
+            continue_nodes, len(new_nodes), replace=False
+        )
+
+        tables = ancient_ts.tables
+        tables.set_samples(continue_nodes, samples=False)
+        tables.union(
+            recent_ts.tables,
+            node_map,
+            add_populations=False,
+            check_shared_equality=False,
+            shift_time=time,
+        )
+
+        return tables.tree_sequence()
+
+    def set_samples(self, nodes, samples=True):
```

Comment:
I don't think we want this method either - it's a very general name for doing something quite specific. If we add methods to the namespace of the TableCollection we need to (a) be very sure that the operation we're adding is generally useful, and not blocking us from doing something similar in the future under this name (i.e., is there something slightly different, but incompatible we might want to do in future under the same name?); (b) make sure it's very thoroughly tested and documented.

Since all we're doing is updating the flags, it would be much simpler to do something like
```python
tables.nodes.flags = new_flags
```